telemetry/firmware/telemetry.c

```diff
@@ -0,0 +1,353 @@
+        queue_UART_push(UART_queue, &(can_recv_msg[0]));
+        queue_UART_push(UART_queue, &(can_recv_msg[1])); 
+        queue_UART_push(UART_queue, &(can_recv_msg[2]));
+        queue_UART_push(UART_queue, &(can_recv_msg[3]));
+        queue_UART_push(UART_queue, &(can_recv_msg[4]));
+        queue_UART_push(UART_queue, &(can_recv_msg[5]));
+        queue_UART_push(UART_queue, &(can_recv_msg[6])); 
+        queue_UART_push(UART_queue, &(can_recv_msg[7])); 
+
+        //Setup to Receive Again for future
+		CANSTMOB = 0x00;
+	    CAN_wait_on_receive(CELL_VOLT_TEMP_MBOX, CAN_ID_BMS_TEMP_12, CAN_LEN_BMS_TEMP_12, CAN_MSK_6);
+    }
+}
+
+ISR(USART_TX_vect){
+    gFlag |= _BV(UART_SEND_READY); // set flag and get back to it 
+}
+
+
+
+/*----- Functions -----*/
+
+
+void initUART(void){
+    // setup UART connection at 9600 8N1 for starters to communicate with RF module 
+    LINCR |= _BV(ENABLE_LIN_AND_UART) |= _BV(ENABLE_UART) |= _BV(ENABLE_RX) |= _BV(ENABLE_TX); 
+
+    // LDIV[11..0] = (FCLK / LBT[5..0] * UART_BAUD)-1; // setup baud rate, pg 183 - is this access of bits within register ok? 
+    LINBRR = 0x0C; // set buad to 9600
+
+    // enable recieve, transmit UART interrupt
+
+    // LINENIR |= _BV(LENRXOK); 
+    LINENIR |= _BV(LENTXOK); 
+}
+
+
+void send_UART(void){ // called from flag check in superloop 
+    if (queue_UART_count(&UART_queue)>0){
+        // pop byte off UART_queue and write to data register 
+        uint8_t data_byte; 
+        if (queue_UART_pop(&UART_queue, &data_byte)==0){
+            LINDAT = data_byte;                 
+            gFlag &= ~_BV(UART_SEND_READY); 
+        } else {
+            // unsuccessful data dequeue from UART_queue
+        }
+    }
+}
+
+/*
+Let's handle data pass back from RF module 
+*/
+void handle_UART_recieved(void){
+}
+
+/*----- MAIN -----*/
+
+int main(void){
+
+    // inits 
+    initUART(); 
+
+	CAN_init(CAN_ENABLED);
+
+    // setup mbox recieve
+	CAN_wait_on_receive(THROTTLE_CMD_MBOX, CAN_ID_MC_COMMAND, CAN_LEN_MC_COMMAND, CAN_MSK_SINGLE);
+	CAN_wait_on_receive(THROTTLE_BRD_MBOX, CAN_ID_THROTTLE, CAN_LEN_THROTTLE, CAN_MSK_SINGLE);
+	CAN_wait_on_receive(BRAKE_MBOX, CAN_ID_BRAKE_LIGHT, CAN_LEN_BRAKE_LIGHT, CAN_MSK_SINGLE);
+	CAN_wait_on_receive(AIR_CONTROL_SENSE_MBOX, CAN_ID_AIR_CONTROL_SENSE, CAN_LEN_AIR_CONTROL_SENSE, CAN_MSK_SINGLE);
+	CAN_wait_on_receive(BMS_CORE_MBOX, CAN_ID_BMS_CORE, CAN_LEN_BMS_CORE, CAN_MSK_SINGLE);
+
+	CAN_wait_on_receive(CELL_VOLT_TEMP_MBOX, CAN_ID_BMS_TEMP_12, CAN_LEN_BMS_TEMP_12, CAN_MSK_6);
+
+
+    queue_CAN_init(&CAN_queue); 
+    queue_UART_init(&UART_queue); 
+
+
+
+    // super loop 
+    while (1) {
+        if (bit_is_set(gFlag, UART_SEND_READY)){
+            send_UART(); 
+        }
+        if (bit_is_set(gFlag, UART_RECIEVED)){
+            handle_UART_recieved();
+        }
+        if (shutdown_error_flag){
+            // TODO: need a clean way to handle this HIGH PRIORITY message & send immediately, without enqueueing? 
```

Comment:
Not sure if we do. If we think about the use case, it would be best to see what the CAN messages _right before_ shutdown are, and if we bypass the queue, we lose ordering. I think we should still use the queue. Shutdown shouldn't stop this board from working, so we'll continue transmitting telemetry data after shutdown.